routes/cube_routes.js

```diff
@@ -1248,103 +1203,72 @@ router.post('/bulkuploadfile/:id', ensureAuth, async (req, res) => {
   }
 });
 
-async function bulkuploadCSV(req, res, cards, cube) {
-  let added = [];
-  let missing = '';
-  let changelog = '';
-  for (let card_raw of cards) {
-    let split = util.CSVtoArray(card_raw);
-    let name = split[0];
-    let maybeboard = split[8];
-    let card = {
-      name: name,
-      cmc: split[1],
-      type_line: split[2].replace('-', '—'),
-      colors: split[3].split('').filter((c) => [...'WUBRG'].includes(c)),
-      set: split[4].toUpperCase(),
-      addedTmsp: new Date(),
-      collector_number: split[5],
-      status: split[6],
-      finish: split[7],
-      imgUrl: split[9] && split[9] != 'undefined' ? split[9] : null,
-      tags: split[10] && split[10].length > 0 ? split[10].split(',') : [],
-    };
+router.post('/bulkreplacefile/:id', ensureAuth, async function(req, res) {
+  if (!req.files) {
+    req.flash('danger', 'Please attach a file');
+    return res.redirect('/cube/list/' + req.params.id);
+  }
 
-    let potentialIds = carddb.allIds(card);
-    if (potentialIds && potentialIds.length > 0) {
-      // First, try to find the correct set.
-      let matchingSet = potentialIds.find((id) => carddb.cardFromId(id).set.toUpperCase() == card.set);
-      let nonPromo = potentialIds.find(carddb.notPromoOrDigitalId);
-      let first = potentialIds[0];
-      card.cardID = matchingSet || nonPromo || first;
-      if (maybeboard == 'true') {
-        cube.maybe.push(card);
+  const items = req.files.document.data.toString('utf8'); // the uploaded file object
+
+  try {
+    const cube = await Cube.findOne(build_id_query(req.params.id));
+    if (!cube) {
+      req.flash('danger', 'Cube not found');
+      return res.status(404).render('misc/404', {});
+    } else if (cube.owner != req.user._id) {
+      req.flash('danger', 'Not Authorized');
+      return res.redirect('/cube/list/' + req.params.id);
+    }
+    cards = items.match(/[^\r\n]+/g);
+    if (cards) {
+      if (cards[0].trim() == CSV_HEADER) {
+        cards.splice(0, 1);
+        try {
+          const { newCards, newMaybe, missing } = CSVtoCards(cards, carddb);
+          const { only_a, only_b, in_both, all_cards } = await compareCubes(
+            req,
+            res,
+            cube,
+            {
+              cards: newCards,
+              maybe: newMaybe,
+            },
+            carddb,
+            prices,
+          );
+          let changelog = '';
+          only_a.forEach((card) => (changelog += removeCardHtml(carddb.cardFromId(card.cardID))));
+          only_b.forEach((card) => (changelog += addCardHtml(carddb.cardFromId(card.cardID))));
+          cube.cards = newCards;
+          cube.maybe = newMaybe;
+          await generateBlogpost(req, res, cube, changelog, only_b.concat(newMaybe), missing, carddb);
```

Comment:
In general I'm not a fan of passing request around, but defer to consensus.
I'd pass the info that the function needs (req.param.id) and throw errors. This means the route controller handles the request failures and the unit tests can be simpler. The flash message then becomes the error message.
Additionally if all errors are thrown and caught in the outer try/catch then there is a single place to flash and redirect.

Reply:
I think this is something that would probably be better taken up as a separate task to restructure the code to whatever we think is best. I do agree that not passing them around would likely be better.